radicle/src/cob.rs

```diff
@@ -0,0 +1,121 @@
+pub use cob::{
+    identity, object::collaboration::error, CollaborativeObject, Create, Entry, History, ObjectId,
+    TypeName, Update,
+};
+use radicle_cob as cob;
+use radicle_git_ext::Oid;
+
+use crate::{
+    identity::{project::Identity, Did},
+    storage::git::Repository,
+};
+
+/// The `Author` of a [`create`] or [`update`].
+///
+/// **Note**: `Author` implements [`identity::Identity`], but since it
+/// is not content-addressed, the [`identity::Identity::content_id`]
+/// returns [`git2::Oid::zero`]. This means that if the `author` is
+/// set in the updates the history entries for those changes will
+/// contain the zero `Oid` for the `author` field.
+pub struct Author {
+    did: Did,
+}
+
+impl identity::Identity for Author {
+    type Identifier = String;
+
+    fn is_delegate(&self, delegation: &crypto::PublicKey) -> bool {
+        *self.did == *delegation
+    }
+
+    fn content_id(&self) -> Oid {
+        git2::Oid::zero().into()
+    }
+
+    fn identifier(&self) -> Self::Identifier {
+        self.did.encode()
+    }
+}
+
+/// Create a new [`CollaborativeObject`].
+///
+/// The `repository` is the project this collaborative object is being
+/// stored under.
+///
+/// The `signer` is used to cryptographically sign the changes made
+/// for this update. **Note** that the public key for the signer must
+/// match the key of the `Author` -- if it is set.
+///
+/// The `project` is used to store its content-address in the history
+/// of changes for the collaborative object.
+///
+/// The `args` are the metadata for this [`CollaborativeObject`]
+/// udpate. See [`Update`] for further information.
+pub fn create<S>(
+    repository: &Repository,
+    signer: S,
+    project: &Identity<Oid>,
+    args: Create<Author>,
+) -> Result<CollaborativeObject, error::Create>
+where
+    S: crypto::Signer,
+{
+    let namespace = *signer.public_key();
+    cob::create(repository, signer, project, &namespace, args)
```

Comment:
Would it make sense for the `cob::update/create` functions to infer the namespace from the signer? It would at least prevent eg. `cob::create(repo, bob_signer, proj, alice_namespace, args)`

Reply:
I didn't want to do that because it being the namespace is specific to this particular storage layout. `radicle-cob` knowing about that doesn't seem clean to me. Otherwise, we could just write `radicle-cob` knowing the exact layout :shrug:

Reply:
I see I see hmmmm. let's leave it as is :+1: